system/lib/compiler-rt/wasm_setjmp.c

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2024 The Emscripten Authors.  All rights reserved.
+ * Emscripten is available under two separate licenses, the MIT license and the
+ * University of Illinois/NCSA Open Source License.  Both these licenses can be
+ * found in the LICENSE file.
+ *
+ * a runtime implementation for
+ * https://github.com/llvm/llvm-project/pull/84137
+ * https://docs.google.com/document/d/1ZvTPT36K5jjiedF8MCXbEmYjULJjI723aOAks1IdLLg/edit
+ *
+ * a plan is to replace the __USING_WASM_SJLJ__ part of emscripten_setjmp.c
+ * with this.
+ */
+
+#ifdef __USING_WASM_SJLJ__
+
+#include <stddef.h>
+#include <stdint.h>
+
+/*
+ * jmp_buf should have large enough size and alignment to contain
+ * this structure.
+ */
+struct jmp_buf_impl {
+        void *func_invocation_id;
+        uint32_t label;
+
+        /*
+         * this is a temorary storage used by the communication between
+         * __wasm_sjlj_longjmp and WebAssemblyLowerEmscriptenEHSjL-generated
+         * logic.
+         * ideally, this can be replaced with multivalue.
+         */
+        struct arg {
+                void *env;
+                int val;
+        } arg;
+};
+
+void
+__wasm_sjlj_setjmp(void *env, uint32_t label, void *func_invocation_id)
+{
+        struct jmp_buf_impl *jb = env;
```

Comment:
Should we just declare the first argument as `struct jmp_buf_impl *jb`? (i.e. why case from void* here?)

Reply:
i don't want to have jmp_buf_impl in the function prototype as it's an internal detail.

Reply:
Fair enough.   I was thinking that because this function is not actually declared anywhere else then it would not matter since this internal detail would remains solely within this `.c` file.